scalameta/common/shared/src/main/scala/scala/meta/internal/trees/ast.scala

```diff
@@ -336,7 +350,54 @@ class AstNamerMacros(val c: Context) extends AstReflection with CommonNamerMacro
           }
           internal(...$internalArgss)
         }
-      """
+        """
+
+        // step 13a: generate additional binary compat Companion.apply
+        // generate new applies for each new field added
+        // with field A, B and additional binary compat ones C, D and E, we generate:
+        // apply(A, B, C), apply(A, B, C, D), apply(A, B, C, D, E)
+        val sortedBinaryCompatName = binaryCompatVars
+          .collect { case vr: ValDef => vr }
+          .sortBy { vr =>
+            try {
+              vr.mods.annotations.collectFirst { case q"new binaryCompatField($version)" =>
+                val since = version.toString().stripPrefix("\"").stripSuffix("\"")
+                val versions = since.split("\\.").map(_.toInt)
+                (versions(0), versions(1), versions(2))
```

Comment:
Does this provide a stable sort when we have two bincompat fields with the same name?

Reply:
`sortBy` is stable judging by the docs so this should also be stable. So if we ever need to add two elements at the same time it should work ok.